vtex/go-io
# Refactor Redis pub/sub connection handling

#### What is the purpose of this pull request?
To improve reliability of the "subcription connection" management code,
by isolating it into a separate component solely responsible for maintaining
a healthy connection with a list of patterns properly subscribed.

This, together with #14, breaks the single lock we had for everything into
individual locks for the connection, each of the subscriptions and the maps
of patterns into subscriptions to support the pub/sub multiplexing.

#### What problem is this solving?
Management of pub/**sub** connections in `redis` package.

Also 2 parallel improvements performed here:
 - Update `redigo` library to it's latest version and use a `ReadTimeout`
   on connections to support having a timeout for Redis commands (I
   believe we were potentially waiting forever before).
 - Properly handle `pong` messages from Redis, as before we only sent
   the `ping`s but didn't do anything about the pong or have any guarantees
   on whether the connection was healthy (i.e. pongs arriving back).

I could make these changes in another separate PR but thought it was not
really necessary. If desired I can do it though, so just let me know.

#### How should this be manually tested?
 - Launch a beta colossus
 - Check pub/sub works properly without losing clients or reaching 
   deadlocks
 - Release colossus
 - Monitor number of go-routines to check the leak is gone (prometheus) 
 - Monitor Redis errors logged to Splunk and make sure nothing new 
   has arosen (like the `pong` never being received)

#### Types of changes
